### DISCUSSION TITLE
nuclei multi threading

### DISCUSSION THREAD

**Discussion by molitona** [projectdiscovery/nuclei#3534] [Q&A]
Hi, why isn't there -t option for multi threading like httpx have ?

nuclei is slow compared to httpx

**Comment by olearycrew** [comment]
@ehsandeep what are your thought here?

**Comment by tarunKoyalwar** [comment]
@molitona  https://nuclei.projectdiscovery.io/nuclei/get-started/#mass-scanning-using-nuclei

**Comment by molitona** [comment]
Hi @tarunKoyalwar @olearycrew @ehsandeep 

I scan 16 million list with only 1 template, that's my case, is below command the right one for my case for faster work ?

```
nuclei - -l 16_milion_list -id xxxx -stats -rl 400 -bs 400 
```

NB: that dash after nuclei word is steam according to https://nuclei.projectdiscovery.io/nuclei/get-started/#mass-scanning-using-nuclei 'cause i saw ---> ```This option should only be enabled if targets > 10k . This skips any type of sorting or preprocessing on target list.``` on that link and it's my case i think ?

i run on vps machine with 500gb speed.

can you suggest the perfect tuned command args for the above one or it's perfect ? also is that steam dash correct syntax to put it after nuclei keyword ?

**Comment by tarunKoyalwar** [comment]
@molitona , interesting and thanks for sharing detailed information this is helpful . the ideal command for your setup should be

also its not a dash we will replace it with `NA` , it means it does not have short flag

```console
nuclei -rl 0 -bs 10000 -l 16_milion_list -id xxx -stats -elog errors.txt -stream -o output.txt -nh
```
i set the `-bs` value to `10000` but i would recommend to test it for more (i think you can even set its value to `50000` or even `100000` if you have enough RAM)

### Some Explaination
`-rl 0` : you are only running 1 template so no need of ratelimit (anything other than 0 will affect your scan speed since ratelimit is global)
`-bs 10000`: also known as host concurrency (this is parameter you should tune for faster scans)
`-stream`  : skips preprocessing like sorting etc
`-nh` :  we can skip this if we are only sending one request per target (if more than 1 not using the flag is more helpful)


PS: I would like to hear back from you about stats like scan time etc if possible

**Comment by molitona** [comment]
Hi @tarunKoyalwar,

Thank you for the detailed response.

I have 4GB RAM. So what is the max -bs value that i don't have to exceed according to your experience ?


Other than that i'll get back to you once I run a test.

Again, thanks for help.

**Comment by tarunKoyalwar** [comment]
@molitona  , if its 4GB i think you might have to drop down bs to `-bs 5000` . this is my estimation but you will get a better understanding after doing 1 or 2 scans whether to increase or decrease concurrency (just keep an eye out on RAM consumption and RPS in stats)

**Comment by molitona** [comment]
that was the case. what about 1gb ram ? how you calculate it ?

Apr 15, 2023, 16:31 by ***@***.***:

**Comment by dogancanbakir** [comment]
This discussion closed automatically due to inactivity. Feel free to reopen or start new if still relevant.
